### ISSUE TITLE
Support transient state in typed persistent actors

### ISSUE THREAD

**Issue by s4nk** [akka/akka#25518]
This is a follow up of the discussion here: https://gitter.im/akka/dev?at=5b742677e9b96f2c98104c98

As described in https://discuss.lightbend.com/t/creating-children-actors-in-typed-persistent-actors/1743 working with transient state in persistent typed actors is not ideal at the moment. As suggested by @patriknw in the discussion above, the simplest available solution is to wrap the persistent behavior in a `Behavior.setup`, create the transient state there and use/modify it from the persistent behavior. It works but requires the transient state to be mutable, which doesn't feel in line with the approach prevailing in Akka Typed, of keeping state immutable and changing the behavior instead. The current persistent typed API is great because it ensures that the persistent state stays consistent and reproducible. Perhaps it would be possible to change it in a way that would make working with transient state more idiomatic without loosing the existing guard rails. Here's how it could look like:
```
object PersistentBehaviors {

  type CommandHandler[Command, Event, PersistentState, TransientState] = (ActorContext[Command], PersistentState, TransientState, Command)
    ⇒ Effect[Event, PersistentState, TransientState]

  type EventHandler[PersistentState, TransientState, Event] = (PersistentState, TransientState, Event) ⇒ (PersistentState, TransientState)

  def receive[Command, Event, PersistentState, TransientState](
    persistenceId: String,
    emptyPersistentState: PersistentState,
    emptyTransientState: TransientState,
    commandHandler: CommandHandler[Command, Event, PersistentState, TransientState],
    eventHandler: EventHandler[PersistentState, TransientState, Event]): PersistentBehavior[Command, Event,
    PersistentState, TransientState] =
    PersistentBehaviorImpl(persistenceId, emptyPersistentState, emptyTransientState, commandHandler, eventHandler)
}

trait PersistentBehavior[Command, Event, PersistentState, TransientState] extends DeferredBehavior[Command] {
  def onRecoveryCompleted(callback: (ActorContext[Command], PersistentState, TransientState) ⇒ TransientState): PersistentBehavior[Command, Event, PersistentState, TransientState]
}
```
And an example of using it:
```
case class PersistentState(items: Seq[Item])

case class TransientState(children: Seq[ActorRef[Child.Command]])

PersistentBehaviors
  .receive[Command, Event, PersistentState, TransientState](
    persistenceId = "my-persistent-actor",
    emptyPersistentState = PersistentState(Seq.empty),
    emptyTransientState = TransientState(Seq.empty),
    commandHandler = commandHandler,
    eventHandler = ???
  )
  .onRecoveryCompleted { (ctx, persistentState, transientState) =>
    val children = persistentState.items.map(item => ctx.spawn(Child(item), item.name))
    TransientState(children)
  }

  def commandHandler: CommandHandler[Command, Event, PersistentState, TransientState] = { (ctx, persistentState, transientState, command) =>
    command match {
      case CommandTriggeringChildCreation(item) =>
        Effect
          .persist(NewItemEvent(item))
          .andThen { (persistentState, transientState) =>
            transientState.copy(children = ctx.spawn(Child(item), item.name) +: transientState.children)
          }
      ...
    }
  }
```

**Comment by patriknw** [comment]
Thanks for creating the issue.

We are actually considering adding an enclosing class because the type signatures becomes rather tedious to repeat. You can follow that in issue #25484 and a first prototype in https://github.com/akka/akka/pull/25503 and feel free to chime in. I would guess that your immediate reaction will be against, but let's collect pros and cons.

**Comment by IainHull** [comment]
The motivation for this is that we are using a persistent actor to model a job queue.  Each tenant has their own job queue which is modelled by a persistent actor.  Once a job is activated we create a child actor (this actor is responsible for monitoring the execution of the job and ensuring it moves to a completed state). When the job is completed we no longer need the child actor.

We do not want to create and destroy loads of job actors during recovery. ESP since they might trigger recovery of the job on the remote host that is executing it.

**Comment by patriknw** [comment]
Right, it's a good use case for this. You would re-create the actors in `onRecoveryCompleted` based on the current (persistent) state.

**Comment by patriknw** [comment]
BTW, for child actors you might not have to maintain them in your own transient state since you could give them a name and use `ActorContext.child(name)` to lookup.

**Comment by rkuhn** [comment]
The provided use-case is a red herring: no actors need to be created during replay, the currently needed list of actors should be tracked in the State; then upon recovery completion only the currently needed actors are created.

In general, I see no possibility to distinguish between so-called persistent and transient state, what would the criterion be? State is always ephemeral, it is a projection of the event log, defined by the currently deployed event handler function and its initial value. If no distinction can be made, why would there be separate APIs?

**Comment by rkuhn** [comment]
Another remark: everything that is an input to command validation is simply the State of the actor—“persistent” or “transient” does not play any role here.

**Comment by IainHull** [comment]
@rkuhn I agree with your assertion of not distinguishing between so-called persistent state and transient state. But if that is the case `onRecoveryCompleted` should have the option to modify the state. 

The idea of keeping them separate was to keep it inline with #23848:

> onRecoveryComplete should not allow for modification of the state, only persisting and replaying events should allow for state modifications

Some state is expensive to create, so PersistentActors should have the option to only create the state they required after recovery.

We are discovering another state requirement where we are holding actor references to non-persistent actors inside a persistent actor. After recovery these should be quarantined as we need verify that these references are valid before we can use them. These types of design would be feasible if `onRecoveryComplete` was able to change the state as opposed to being purely side effecting.

**Comment by rkuhn** [comment]
When discussing changes to the fundamental type signatures of the package we need to always keep in mind the fundamental principle on which the package is built. In the case of Akka Persistence the principle is that we guarantee the recovery of a Persistent Actor with exactly the State it had reached before the crash by replaying the event log (if the code has not changed in the meantime). Anything that breaks this guarantee needs an equally fundamental reason—with a strong bias towards not breaking existing designs based upon it.

In the following we shall consider only the case that the code does not change after a crash; changing the code is a necessary but different feature. This means that event handler and initial state value are assumed to be stable across crashes.

We currently achieve the guarantee by restricting state changes to exactly those recorded in the event log—any other means of changing the state would break it. It follows that allowing a state change upon completed recovery is ruled out unless that change is mediated by a recorded event; the onRecoveryComplete hook would not directly emit a new state, it would emit an additional event.

The use-case you highlight is indeed an interesting one: if the computation of state is sufficiently complex so that replay of a log of expected length takes longer than tolerable, one solution would be to defer the computation until all events have been seen—this helps under the assumption that there is a batching effect or that parts of the computation can be elided since its consequences have been canceled by later events.

Another solution is to store snapshots in order to reduce the number of events to be replayed. This solution has the advantage of providing an upper bound on recovery time, which the first proposal does not: accumulating facts from an unbounded event log may take arbitrarily long.

On the topic of ActorRefs: if the persistent actor needs to know their readiness it will have to employ DeathWatch on them in any case, which can easily be done from `onRecoveryComplete`. Then, dead references are weeded out by handling termination notifications and emitting corresponding removal or refresh events.

---

In general, immutable state and referential transparency only have benefit if they are kept alive through the whole design. Allowing “nice looking” features into the codebase to save a few characters will render them useless.

People of course reach to familiar tools first, which for some audiences might be completely absent in the Akka Typed world. But I posit that enough change towards functional programming and immutability has trickled through the Java ecosystem over past decade. We only need to provide answers in the form of well-documented patterns to move this safer way of programming fully into the familiar zone.

**Comment by patriknw** [comment]
The event replay should be free of side effects, ofc. Therefore you can't for example spawn actors that are performing some tasks during the even replay. Some of these tasks may already have been completed and that is recorded by a later event. A good place to spawn remaining tasks would be in onRecoveryComplete based on the recovered State.

The question here was if the ActorRefs of these tasks would belong in the State or be on the side. The ActorRefs can't be persisted so if inside the State they would be transient fields. Where can the state be updated with the new ActorRefs?  Emitting another event like `TasksSpawnedAgain` from `onRecoveryComplete` feels a bit redundant because we have already recorded that the tasks have started.

There are other examples of this. Like resending unconfirmed messages and keeping track of resend counts, backoff durations and such, without having to store an event for every such transient state change.

**Comment by patriknw** [comment]
Here we talked about storing child ActorRefs. No point since they are always terminated when recovering new parent.

Ok, good point that counts can be kept in timer messages.

To be clear, I don’t see the need for a change for this but I want to find the right advice, to be documented. I’m not against mutable transient state in setup if needed.

**Comment by s4nk** [comment]
@rkuhn I totally agree that breaking the principles behind `PersistentBehavior` is a bad idea, and that the decorator approach is preferable. 

Thinking about examples - two types of state come to mind:
- state which becomes obsolete when an actor dies (e.g., ActorRefs of its child actors)
- state which is derived from commands but recording it in the journal would be undesirable because of performance or size (it's hard to come up with a reasonable example but I'll try :) - an in-memory blob cache, where commands are instructions to add or remove blobs specified by URIs; of course the events should only contain URIs, but the blobs need to be kept somewhere; the only options are to wrap the PersistentBehavior and use a `var` or delegate the blob management to a non-persistent actor, but that feels like adapting domain design to the framework constraints)

**Comment by rkuhn** [comment]
There is a far simpler solution for the blob cache: this responsibility should clearly be owned by a non-persistent actor, since the cache is not meant to be persistent. This should BTW be considered far more often, delegating to other actors is at the core of the actor model.

For a persistent cache the blobs will need to be stored in the event log since otherwise the actor is not guaranteed to recover to the same state: the data behind a URL are not immutable. Given the implied cost and the usual purpose of caches I think it unlikely that such a case be made. Looking at it from the other side, this also feels more like an archive than a cache.
